### PR TITLE
fix(grammar): close 13 syntax gaps (CRLF, quoting, expansions, heredocs, image refs)

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

Closes a batch of syntax gaps found in a deep audit of the grammar. Each fix was validated against real `docker build` 29.6.1 / BuildKit behavior and independently re-verified against the live parser.

**Corpus:** 161 → **183 tests, all passing** · Go tests pass · all 9 `examples/` parse clean · no regressions found across a broad real-world idiom sweep.

## Fixed (13 gaps)

| Gap | Before | After |
|---|---|---|
| **CRLF continuations** | `RUN a \` + CRLF + `b` → `ERROR` (every Windows/autocrlf multi-line RUN) | valid `line_continuation` |
| **Adjacent value segments** | `LABEL k=pre"mid"post` → `ERROR` on 2nd segment | segments concatenate (shared `adjacentValue()` helper) |
| **Literal `$`** | `ENV PRICE=$5`, `cost:$5.00`, `$$` → `ERROR` | literal text; real `$var`/`${...}` still expand |
| **Registry port** | `registry:5000/img:tag` → port+path swallowed into tag | `name=registry:5000/img`, `tag=tag` |
| **USER / stage names** | `USER _apt`, `USER my.user`, `AS my.stage` → `ERROR` | parse cleanly |
| **EXPOSE** | `80/sctp`, `80/TCP`, `$PORT/tcp` → `ERROR` | accepted |
| **End-of-options** | `COPY -- src dst` → `ERROR` | `end_of_options` node |
| **ENV legacy quotes** | `ENV GREETING hello "world"` → `ERROR` | quote is literal rest-of-line (matches LABEL) |
| **Quote escaping** | single-quoted Windows path ending in a backslash → `MISSING` quote; escaped `\$` mislabeled as a live expansion | single quotes fully literal; escaped `\$` is a literal `escape_sequence` |
| **Expansion modifiers** | `${var:-default}` folded operator+word into the variable name | `variable` + `expansion_modifier` (`expansion_operator`); nested `${foo:-${bar}}` too |
| **Flag `$`-vars** | `--platform=$BUILDPLATFORM`, `--chown=$U:$G` opaque | surface `expansion` nodes |
| **Heredoc body / injection** | injected JSON/YAML/bash ingested the `EOF` delimiter line | named `heredoc_content` node; injections retargeted to it |

Also corrected 3 pre-existing corpus tests that asserted behavior real Docker **rejects** (backslash/backtick escaping a closing single quote).

## Not fixed — documented tradeoffs (all pre-existing, none are regressions)

- **Instruction keyword word-boundary** (`FROMalpine` mis-parses): tree-sitter's regex engine supports no `\b`/lookahead.
- **Mid-path continuation** (`WORKDIR /app/\` + newline + `bin`): cannot be distinguished from the common space-separated multi-path case without an immediate-only continuation token; kept the common `VOLUME`/`COPY` case safe.
- **`--mount` option values** kept as a complete literal run (no `$`-expansion node): interleaving literal+expansion in the comma-separated mount grammar truncated the value, so it stays opaque to guarantee complete capture (matches prior behavior). Generic `--flag` values still surface expansions.
- Leading-quote-plus-trailing-content in ENV space-form, and dash-prefixed paths after `--`, still error — these also error on `main`.

## ⚠️ Before merge

1. **ABI version:** commits were generated with tree-sitter CLI 0.26.9 (ABI 15) and the branch tip normalized to **ABI 14** to match the repo pin (0.26.10). Please **squash-merge**, or run `tree-sitter generate --abi 14` before merging, so `src/parser.c` stays ABI 14.
2. A stale `containerfile.dylib` build artifact in the repo root can desync `tree-sitter parse` from the grammar during local dev — worth `.gitignore`-ing separately.

## Test plan

- `tree-sitter test` → 183/183
- `go test ./...` → pass (run `go clean -cache` first after regenerating)
- Every `examples/*` parses with no `ERROR`/`MISSING`
- Each fix cross-checked against `docker build --check` / full builds on Docker 29.6.1

## Summary by CodeRabbit

* **New Features**
  * Enhanced parsing for quoted/unquoted value concatenation, expansion operators/modifiers, and more consistent heredoc content structure.
  * Improved `ADD`/`COPY`/`--` handling (including dash-prefixed paths) and `EXPOSE` protocol validation.
  * Extended `USER`/group name parsing to allow dots; improved highlighting for expansion operators and heredoc-only content injection.
* **Bug Fixes**
  * Fixed CRLF line-continuation edge cases and improved literal `$` handling in parameter/mount values.
* **Tests**
  * Expanded corpus coverage for `ARG`, `ENV`, `COPY`/`ADD`, `EXPOSE`, heredocs, and related quoting scenarios.
* **Chores**
  * Updated CRLF fixture handling and adjusted Go CI cache behavior.

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -222,10 +281,12 @@ export default grammar({
...
-                $.single_quoted_string,
-                $.unquoted_string,
+              adjacentValue(
+                $,
+                choice(
+                  $.double_quoted_string,
+                  $.single_quoted_string,
+                ),
               )),
           ),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:317]: Good catch — fixed in 523d3c9. Wrapped the `default` field in `optional(...)` so `ARG MY_VAR=` (empty default) parses. Verified against `docker build` and added a corpus test ("Empty default value"). Note: empty `ENV K=` and `LABEL k=` already parsed correctly (their value fields were structured differently), so this was ARG-specific.

File: grammar.js [hunk 2]
```diff
@@ -422,17 +516,25 @@ export default grammar({
...
-            $.double_quoted_string,
-            $.single_quoted_string,
-            $.unquoted_string,
+          adjacentValue(
+            $,
+            choice(
+              $.double_quoted_string,
+              $.single_quoted_string,
+            ),
           )),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:590]: Thanks — I checked this specific case and empty `LABEL my_label=` already parses cleanly on this branch (the LABEL value path differs from ARG). I verified `LABEL k=` produces no ERROR node, so no change was needed here. The ARG empty-default sibling issue was real and is fixed in 523d3c9.

File: grammar.js [hunk 3]
```diff
@@ -467,7 +569,16 @@ export default grammar({
...
+        repeat(
+          choice(
+            token.immediate(/[^@:\s\$]+/),
+            // A registry host may include a port: the `:PORT/` before the
+            // repository path is part of the name, not the tag (which is the
+            // colon after the final path segment).
+            token.immediate(/:\d+\//),
+            $._immediate_expansion,
+          ),
+        ),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:635]: Fixed in 523d3c9. Extended the port token to accept an expansion: `token.immediate(/:(\d+|\$[a-zA-Z_][a-zA-Z0-9_]*|\$\{[^}]+\})\//)`. I kept it as a single token (rather than `seq(":", expansion, "/")`) on purpose — the multi-symbol form created an unresolved LR conflict with `image_tag`'s leading `:`. `FROM registry:${PORT}/myimage` and `FROM registry:$PORT/img` now parse cleanly (verified `${PORT}` expands under `docker build`), with a corpus test added.

File: grammar.js [hunk 4]
```diff
@@ -374,11 +441,31 @@ export default grammar({
...
+          token.immediate(/:?[-+?]/),
+          $.expansion_operator,
+        ),
+        repeat(
+          choice(
+            token.immediate(/[^}$]+/),
+            $._immediate_expansion,
+          ),
+        ),
+      ),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:514]: Fixed in 523d3c9. Added quoted-string alternatives (and excluded `"`/`'` from the literal run) in `expansion_modifier`, so a quoted word containing `}` or `$` no longer terminates the expansion early. `${x:-"blue}"}` now parses with the quoted default as a `double_quoted_string` node; corpus test added and verified against `docker build`.

File: grammar.js [hunk 5]
```diff
@@ -129,6 +186,7 @@ export default grammar({
       seq(
         alias(/[aA][dD][dD]/, 'ADD'),
         repeat($.param),
+        optional(seq($.end_of_options, $._non_newline_whitespace)),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js]: Fixed in 523d3c9, and went further than the suggestion. I dropped the required `_non_newline_whitespace` after `end_of_options` (whitespace/continuations are handled by `extras`, as you note), and — since the whole point of `--` is to allow dash-prefixed source paths — added a `path_with_heredoc_or_dash` operand form used after `--`. So `ADD -- -archive.tar /dest/` and `ADD -- \` + continuation now both parse. Corpus test added.

File: grammar.js [hunk 6]
```diff
@@ -145,6 +203,7 @@ export default grammar({
       seq(
         alias(/[cC][oO][pP][yY]/, 'COPY'),
         repeat($.param),
+        optional(seq($.end_of_options, $._non_newline_whitespace)),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js]: Fixed in 523d3c9 (same change as the ADD sibling): removed the required trailing whitespace and added a dash-allowing operand form after `--`, so `COPY -- \` + continuation and `COPY -- -src /dst` both parse. Corpus test added.

[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js]: Confirmed and fixed in 523d3c9. You're right that the operands were still going through `path_with_heredoc` (which rejects a leading `-`), so `COPY -- -src /dst` errored despite the `--`. Added a post-`end_of_options` operand form (`path_with_heredoc_or_dash`) that permits leading hyphens. `COPY -- -src /dst` and `ADD -- -archive.tar /dst` now parse (verified against `docker build`), with a corpus test.

File: grammar.js [hunk 7]
```diff
@@ -669,6 +790,12 @@ export default grammar({
           token.immediate(/[\\`] /),
           token.immediate(/[\\`][^\s\n]/),
           $._immediate_expansion,
+          // A `$` not starting a valid expansion is literal text. The first
+          // form matches `$` plus the following non-identifier/brace char that
+          // is not a value terminator (handles $5, $$, cost:$5.00); the second
+          // matches a `$` at the end of the value. A real expansion still wins
+          // for $ident and ${...}.
+          token.immediate(/\$[^a-zA-Z_{\s"']/),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js]: Accurate diagnosis, and thanks. I attempted a fix but could not land one cleanly: a trailing `$` is lexically identical to the start of an expansion, and tree-sitter's regex engine has no lookahead/`\b` support, so I cannot write a token that matches `$` *only when* no valid expansion follows without either (a) a standalone `$` token that conflicts with the expansion's `$` (breaks `${...}`), or (b) consuming the following char (which eats the newline and fuses instructions). The common mid-value cases (`$5`, `cost:$5.00`, `$$`) are handled by the `$`+following-char token; only a bare trailing `$` before a newline/EOF (`cost$`, `k=$`) still errors. I've left it as a documented known edge rather than ship a regression. Open to ideas — an external-scanner token could resolve it, but that felt disproportionate for this edge. Not marking resolved.

File: grammar.js [hunk 8]
```diff
@@ -27,6 +27,63 @@ const spacedValueContinuation = ($, atom) =>
...
+    repeat(
+      seq(
+        choice(
+          alias($._immediate_double_quoted_string, $.double_quoted_string),
+          alias($._immediate_single_quoted_string, $.single_quoted_string),
+        ),
+        optional($.unquoted_string),
+      ),
+    ),
+  );
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:91]: Fixed in 523d3c9 — added the blank line after each of the three block descriptions before the `@param`/`@returns` tags. `npx eslint grammar.js` now passes locally, which should clear the `lint` CI check.

File: queries/injections.scm [hunk 9]
```diff
@@ -9,10 +9,12 @@
...
 
+; Inject into the heredoc body only (heredoc_content), which excludes the
+; closing delimiter line so the sub-language does not ingest the terminator.
 ((run_instruction
-  (heredoc_block) @injection.content)
-  (#set! injection.language "bash")
-  (#set! injection.include-children))
+  (heredoc_block
+    (heredoc_content) @injection.content))
+  (#set! injection.language "bash"))
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply queries/injections.scm:17]: I verified this empirically and the newlines are **not** flattened, so I'm keeping `include-children` off.

`heredoc_content` is a single contiguous node whose byte range covers the whole body. The internal `'\n'` tokens are anonymous literal children *within* that span (not gaps between separate named nodes), so the `@injection.content` capture — which uses the node's byte range — already includes them.

Concretely, for:
```
RUN <<EOF
echo one
echo two
echo three
EOF
```
`tree-sitter query queries/injections.scm` reports the bash injection range as `(2,0)`→`(5,0)`, i.e. the exact bytes `echo one\necho two\necho three\n` — all three newlines preserved, and it stops before the `EOF` delimiter line (which was the whole point of switching from `heredoc_block` + `include-children` to `heredoc_content`; the old form pulled the `EOF` line into the injected document).

The general docs note about `include-children` applies when a captured node has named children whose text would otherwise be excluded — that isn't the case for `heredoc_content` (contiguous range, no named children). The Go injection tests (`TestCopyHeredocInjectionsByDestinationExtension`, `TestShellCommandInjectionsAreNotCombined`) also pass. Happy to reconsider if you have a concrete multi-line case that reproduces flattening.

File: grammar.js [hunk 10]
```diff
@@ -483,18 +639,33 @@ export default grammar({
...
+            field('value', $._param_value),
           ),
         ),
       ),
 
+    _param_value: ($) =>
+      repeat1(
+        choice(
+          token.immediate(/[^\s\$]+/),
+          $._immediate_expansion,
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:669]: Confirmed regression and fixed in the just-pushed commit. `_param_value` excluded `$` from its literal arm, so `COPY --exclude=$5` (and `cost:$5.00`, or a value ending in `$`) errored versus the prior `[^\s]+` token. Added the same literal-`$` fallback (`token.immediate(/\$[^a-zA-Z_{\s]/)`) used for unquoted values. `--exclude=$5` and `--foo=cost:$5.00` now parse, while `--platform=$BUILDPLATFORM`/`--chown=$U:$G` still surface expansion nodes. Corpus test added; verified against `docker build`. 👍 useful.

File: grammar.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:91]: Thanks for confirming!

File: queries/injections.scm (hunk 9 above)
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply queries/injections.scm:17]: Appreciate you taking the empirical evidence and withdrawing — thanks!

File: grammar.js [hunk 11]
```diff
@@ -147,6 +217,14 @@ export default grammar({
         repeat($.param),
         choice(
           $.json_string_array,
+          seq(
+            $.end_of_options,
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:225]: Thanks — addressed in the just-pushed commit by adding `(end_of_options)?` after `(param)*` in all four COPY heredoc injection patterns (json/yaml/toml/xml).

One note for transparency: in my testing the injection actually still matched *before* this change, because the patterns have no anchor (`.`) between `(param)*` and `(path (heredoc_marker))`, so an intervening `end_of_options` node was tolerated — `tree-sitter query` on `COPY -- <<EOF config.json` reported both `injection.filename=config.json` and `injection.content`. But making it explicit is clearly better and removes any dependence on that anchoring subtlety, so I applied it. Verified filename+content captures fire for `COPY -- <<EOF x.json`, `COPY --chown=1000 -- <<EOF x.toml`, and the no-`--` baseline. 👍

**Comment by tinovyatkin** [comment]
@codex review

File: grammar.js [hunk 12]
```diff
@@ -669,6 +839,12 @@ export default grammar({
           token.immediate(/[\\`] /),
           token.immediate(/[\\`][^\s\n]/),
           $._immediate_expansion,
+          // A `$` not starting a valid expansion is literal text. The first
+          // form matches `$` plus the following non-identifier/brace char that
+          // is not a value terminator (handles $5, $$, cost:$5.00); the second
+          // matches a bare `$` at the end of a value (cost$, k=$). A real
+          // expansion still wins for $ident and ${...} (longer/immediate).
+          token.immediate(/\$[^a-zA-Z_{\s"']/),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js]: Fixed properly in f6d107b — and thanks for the persistence, you were right that the previous comment overstated the coverage.

I initially tried to solve this with grammar tokens but it is genuinely impossible there: a regex token cannot distinguish a literal `$` from the start of an expansion (tree-sitter's regex engine has no lookahead/`\b`), and any token that absorbs a trailing `$` (e.g. `/[^\s]*\$+/`) also steals the `$` from a following `$identifier`, breaking `text$VAR`.

So I added a `_literal_dollar` **external-scanner** token (`src/scanner.c`): it consumes a `$` only when the next char is NOT an identifier char or `{` — the one-char lookahead the regex lexer lacks — and leaves a real `$ident`/`${...}` to the grammar. Wired into both `unquoted_string` and `_param_value`.

Now clean (verified against `docker build`, corpus + go tests pass):
- `ENV PRICE=cost$`, `ARG SUFFIX=$`, `ENV A=$` (bare), `COPY --exclude=bar$` → literal, no ERROR
- `ENV A=bar$BAZ`, `${HOME}`, `$HOME`, `--platform=$BUILDPLATFORM` → still surface expansion nodes
- `ENV A=$` followed by `RUN ...` → two separate instructions (newline not consumed)

👍

File: grammar.js [hunk 13]
```diff
@@ -27,6 +27,66 @@ const spacedValueContinuation = ($, atom) =>
...
+ */
+const doubleQuotedBody = ($) =>
+  repeat(
+    choice(
+      token.immediate(/[^"\n\\`\$]+/),
+      alias($.double_quoted_escape_sequence, $.escape_sequence),
+      token.immediate(/\$\(/),
+      '\\',
+      '`',
+      $._immediate_expansion,
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:45]: Fixed in 6a47043. Added the `_literal_dollar` external-scanner token to `doubleQuotedBody` too, so a `$` not starting a valid expansion is literal inside double quotes: `ENV PRICE="$5"` and `LABEL x="cost$"` now parse cleanly (verified `docker build` accepts them). Real expansions (`"x$HOME"`, `"${VAR}"`) and command substitution (`"$(cmd)"`) are unaffected — the longer `$(` token still wins over the single-`$` scanner token. 👍

File: grammar.js [hunk 14]
```diff
@@ -422,17 +562,25 @@ export default grammar({
     _env_key: ($) =>
       alias(/[a-zA-Z_][a-zA-Z0-9_]*/, $.unquoted_string),
 
-    expose_port: () => seq(/\d+(-\d+)?/, optional(choice('/tcp', '/udp'))),
+    expose_port: ($) =>
+      seq(
+        choice(/\d+(-\d+)?/, $.expansion),
+        // Protocol is optional and case-insensitive; Docker normalizes it and
+        // accepts tcp, udp, and sctp. It must abut the port (no space).
+        optional(token.immediate(/\/[a-zA-Z]+/)),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js]: Good catch — fixed in 6a47043. The protocol suffix is now restricted to `tcp`/`udp`/`sctp` (case-insensitive: `/([tT][cC][pP]|[uU][dD][pP]|[sS][cC][tT][pP])/`) instead of any alphabetic run. `EXPOSE 80/http` and `80/udpz` now surface as errors again (matching `docker build`, which rejects them with "invalid proto"), while `80/tcp`, `80/TCP`, `80/sctp` remain valid. Added a corpus test. 👍

**Comment by tinovyatkin** [comment]
@codex review

File: grammar.js [hunk 15]
```diff
@@ -374,11 +483,38 @@ export default grammar({
...
+    // are pre-release in BuildKit and `#` collides with comment lexing.
+    expansion_modifier: $ =>
+      seq(
+        $.expansion_operator,
+        repeat(
+          choice(
+            // Exclude quote chars from the literal run so a quoted word
+            // (e.g. ${var:-"blue}"}) is parsed as a string and its `}`/`$`
+            // do not terminate the expansion prematurely.
+            token.immediate(/[^}$"']+/),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:505]: I looked into this and it appears to be a false positive — the modifier literal-run token already excludes `$` (and `}`): on the current tip (`6a47043`) it is `token.immediate(/[^}$"']+/)` (grammar.js:505). So `$` is not allowed in the literal run, and a nested expansion does start at the `$` as you describe.

Verified empirically — `ENV R=${foo:-pre${bar}}` parses cleanly into:
```
(expansion
  (variable)            ; foo
  (expansion_modifier
    (expansion_operator) ; :-
    (expansion           ; the nested ${bar}
      (variable))))      ; bar
```
i.e. `pre` is literal text, `${bar}` is a proper nested `expansion`, no ERROR and no leaked `}`. Same for `${a:-x${b}y}`, `${a:-${b}${c}}`, and `${a:-pre$b}` (all parse with nested expansion nodes).

The `$` exclusion this comment recommends is already in place, so I don't think a change is needed here — but happy to reconsider if you have a specific input that still reproduces. 👎 (not reproducible)

**Comment by tinovyatkin** [comment]
@codex review

File: grammar.js [hunk 16]
```diff
@@ -334,10 +421,16 @@ export default grammar({
...
-        repeat(seq($._heredoc_line, '\n')),
+        // The body (everything before the closing delimiter) is wrapped in a
+        // named node so language injection can target exactly the redirected
+        // content, excluding the heredoc_end delimiter line.
+        optional($.heredoc_content),
         $.heredoc_end,
       ),
 
+    heredoc_content: ($) =>
+      repeat1(seq($._heredoc_line, '\n')),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:432]: Good catch, and correct — fixed in the just-pushed commit. I had been running `tree-sitter generate` directly, which skipped the `tree-sitter-go-types` step, so `node_types.go` was stale. Ran the full `npm run generate` (`tree-sitter generate --abi 14 && tree-sitter-go-types`); `node_types.go` now exposes `NodeHeredocContent`, `NodeEndOfOptions`, `NodeExpansionModifier`, and `NodeExpansionOperator`. Corpus + go tests still pass, ABI stays 14. 👍

**Comment by tinovyatkin** [comment]
@codex review

**Comment by tinovyatkin** [comment]
@codex review

File: grammar.js [hunk 17]
```diff
@@ -374,11 +483,38 @@ export default grammar({
...
+        $.expansion_operator,
+        repeat(
+          choice(
+            // Exclude quote chars from the literal run so a quoted word
+            // (e.g. ${var:-"blue}"}) is parsed as a string and its `}`/`$`
+            // do not terminate the expansion prematurely.
+            token.immediate(/[^}$"']+/),
+            alias($._immediate_double_quoted_string, $.double_quoted_string),
+            alias($._immediate_single_quoted_string, $.single_quoted_string),
+            $._immediate_expansion,
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js:508]: Fixed in 37845db. Added `$._literal_dollar` (the external-scanner literal-dollar token) to the `expansion_modifier` word choices, so `${foo:-$5}` and `${foo:-cost$}` now parse cleanly (verified `docker build` accepts them, resolving to empty/literal). Nested defaults (`${foo:-${bar}}`), quoted defaults (`${foo:-"blue}"}`), and plain (`${foo:-def}`) still work. 👍

File: grammar.js [hunk 18]
```diff
@@ -720,20 +907,30 @@ export default grammar({
...
+        $._immediate_expansion,
+      ),
 
-    single_quoted_escape_sequence: () => token.immediate(
-      /[\\`][\\`']/,
+    // The active escape char (\ by default, or ` via `# escape=`) escapes
+    // ", \, `, and $ inside double-quoted strings. A \$ is literal text (the
+    // following identifier is NOT expanded).
+    double_quoted_escape_sequence: () => token.immediate(
+      /[\\`][\\`"$]/,
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply grammar.js]: Fixed in 37845db, thanks — this was a regression from my earlier change. I had added `$` to the double-quoted escape target set, which made `\$` a literal escape in *both* modes; under `# escape=\`` the backslash is literal and `$HOME` must expand, so that was wrong.

Since the active escape char before `$` is mode-dependent and a static grammar token cannot disambiguate it (I tried an external-scanner token, but tree-sitter's internal `\` token shadows it), I reverted the escape target set to `/[\\\`][\\\`"]/` (its long-standing form). Now `"x\$HOME"` keeps the backslash literal and expands `$HOME` in both modes — correct under the backtick directive, and matching the prior behavior under the default one. Verified against `docker build` (backtick mode → `x\` + expansion). 👍

File: queries/injections.scm [hunk 19]
```diff
@@ -9,55 +9,61 @@
...
   (#set! injection.language "bash"))
 
+; Inject into the heredoc body only (heredoc_content), which excludes the
+; closing delimiter line so the sub-language does not ingest the terminator.
 ((run_instruction
-  (heredoc_block) @injection.content)
-  (#set! injection.language "bash")
-  (#set! injection.include-children))
+  (heredoc_block
+    (heredoc_content) @injection.content))
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply queries/injections.scm:16]: I looked into this and it is not reproducible — the newlines are preserved without `injection.include-children`, so I am keeping it off.

`heredoc_content` is a single contiguous node whose byte range spans the whole body; the internal `_heredoc_line`/`\n` tokens are children *within* that same range (not gaps between separate named siblings), so an `@injection.content` capture on it — which uses the node byte range — already includes them.

Empirically, for:
```
RUN <<EOF
echo one
echo two
EOF
```
`tree-sitter query queries/injections.scm` reports the bash `injection.content` range as `(2,0)`→`(4,0)`, i.e. the exact bytes `echo one\necho two\n` — both lines and both newlines, stopping before the `EOF` delimiter line. The injected document is not empty and has no gaps.

This is the same point CodeRabbit raised and then withdrew after seeing this evidence (see the thread on `queries/injections.scm:17`). The `include-children` docs note applies when a captured node has named children whose text would otherwise be excluded; `heredoc_content` has none (contiguous range). The Go injection tests (`TestCopyHeredocInjectionsByDestinationExtension`, `TestShellCommandInjectionsAreNotCombined`) also pass. Happy to reconsider with a concrete multi-line input that reproduces an empty/gapped injection. 👎 (not reproducible)
